### PR TITLE
EVG-14557: User auth over gRPC

### PR DESCRIPTION
**Pull request by julianedwards**
JIRA: https://jira.mongodb.org/browse/EVG-14557

### CODE REVIEW

File: rest/service.go [hunk 1]
```diff
@@ -286,6 +280,7 @@ func (s *Service) addMiddleware() {
 
 func (s *Service) addRoutes() {
 	checkUser := gimlet.NewRequireAuthHandler()
+	depotCheck := NewCertDepotCheckMiddleware(s.Depot == nil)
```
Comment (Kimchelly) [inline rest/service.go:283]: nit: can this be `checkDepot` for consistency?

File: rest/middleware.go [hunk 2]
```diff
@@ -14,6 +14,25 @@ import (
...
+
+// NewCertDepotCheckMiddleware returns an implementation of gimlet.Middleware
+// that returns an error if the internal cert depot is disabled.
+func NewCertDepotCheckMiddleware(disabled bool) *certDepotCheckMiddleware {
+	return &certDepotCheckMiddleware{depotDisabled: disabled}
+}
+
+func (m *certDepotCheckMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	if m.depotDisabled {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(errors.New("certificate depot disabled")))
```
Comment (Kimchelly) [inline rest/middleware.go:29]: Rather than writing a bad request response, can it return something like 403 forbidden?

File: rest/middleware.go [hunk 3]
```diff
@@ -23,7 +42,7 @@ type evgAuthReadLogByIDMiddleware struct {
 // gimlet.Middleware that sends a http request to Evergreen to check if the
 // user is authorized to read the log they are trying to access based on the
 // given ID.
-func NewEvgAuthReadLogByIDMiddleware(sc data.Connector, evgConf *model.EvergreenConfig) gimlet.Middleware {
+func NewEvgAuthReadLogByIDMiddleware(sc data.Connector, evgConf *model.EvergreenConfig) *evgAuthReadLogByIDMiddleware {
```
Comment (Kimchelly) [inline rest/middleware.go:45]: Does this function) and the one below it need to  be exported? It seems like they're only used within `rest`.

File: rest/routes.go [hunk 4]
```diff
@@ -568,8 +568,11 @@ func (s *Service) fetchUserCertKey(rw http.ResponseWriter, r *http.Request) {
 		}
 	}
 
+	if s.Depot == nil {
```
Comment (Kimchelly) [inline rest/routes.go:571]: Is this supposed to do something or is it handled by the depot-checking middleware?

File: rpc/service.go [hunk 5]
```diff
@@ -21,66 +21,65 @@ import (
...
-	if c.CAName == "" {
-		return errors.New("must specify a CA name")
-	}
-	if c.ServiceName == "" {
-		return errors.New("must specify a service name")
-	}
+func (c *AuthConfig) Validate() error {
+	catcher := grip.NewBasicCatcher()
+
+	catcher.AddWhen(c.TLS && c.UserAuth, errors.New("must specify either TLS or user auth, not both"))
```
Comment (Kimchelly) [inline rpc/service.go:37]: Should this also check for `!c.TLS && !c.UserAuth` so we don't end up in a state with no authentication?

**Review by Kimchelly — CHANGES_REQUESTED** [review]
In terms of deploying new agents, is it possible for gRPC clients to send both the TLS credentials _and_ the API headers temporarily so that the rollover is backward compatible (at least until agents have rolled over and cedar is still using TLS)?

File: rpc/service.go (hunk 5 above)
Comment (julianedwards) [reply rpc/service.go:37]: No auth is useful for testing.

File: rest/routes.go (hunk 4 above)
Comment (julianedwards) [reply rest/routes.go]: I forgot to remove it, done.

**Comment by julianedwards** [comment]
I changed the code so that both user auth and TLS can be enabled.

**Review by Kimchelly — APPROVED** [review]

**Pull request merged**